### PR TITLE
Modules

### PR DESCRIPTION
**Pull request by SquidDev**
I guess??

### CODE REVIEW

File: src/Parser.y [hunk 1]
```diff
@@ -99,15 +104,19 @@ Tops :: { [Toplevel Parsed] }
...
+    | external val ident':' Type '=' string    { withPos2 $1 $7 $ ForeignVal (getName $3) (getString $7) $5 }
 
-    | type Var ListE(TyVar)                          { TypeDecl (getL $2) $3 [] }
-    | type Var ListE(TyVar) '=' List1(Ctor, '|')     { TypeDecl (getL $2) $3 $5 }
-    | type Var ListE(TyVar) '=' '|' List1(Ctor, '|') { TypeDecl (getL $2) $3 $6 }
+    | type ident ListE(TyVar)                          { TypeDecl (getName $2) $3 [] }
+    | type ident ListE(TyVar) '=' List1(Ctor, '|')     { TypeDecl (getName $2) $3 $5 }
+    | type ident ListE(TyVar) '=' '|' List1(Ctor, '|') { TypeDecl (getName $2) $3 $6 }
+
+    | module Con '=' Tops end                  { Module (getL $2) $4 }
```
Comment (SquidDev) [inline src/Parser.y]: `module Con '=' begin Tops end`

File: examples/modules.ml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+module Option =
```
Comment (zardyh) [inline examples/modules.ml]: Can you make this have an explicit `begin` token? I'm fine with either
```ocaml
module Option = struct
...
end
```
or
```ocaml
module Option = begin
...
end
```

File: src/Core/Lower.hs [hunk 3]
```diff
@@ -203,16 +203,20 @@ lowerPat pat = case pat of
...
+                                 (k, _) <- makeBigLams ant
+                                 (v,,) <$> lowerType ant <*> (k <$> lowerExpr ex)))
+      <*> lowerProg prg
+lowerProg (TypeDecl (TvName var) _ cons:prg) =
+  (:) <$> (CosType var <$> do
+              for cons $ \case
+                UnitCon (TvName p) (_, t) -> (,) p <$> lowerType t
+                ArgCon (TvName p) _ (_, t) -> (,) p <$> lowerType t)
+      <*> lowerProg prg
+lowerProg (Open _ _:prg) = lowerProg prg
```
Comment (zardyh) [inline src/Core/Lower.hs]: Idea: lower this to a bunch of lets accessing records, I think.

```ocaml
open (Foo : module val foo; val bar end)
(* into *)
let foo = Foo.foo
let bar = Foo.bar
```

File: src/Core/Lower.hs [hunk 4]
```diff
@@ -203,16 +203,20 @@ lowerPat pat = case pat of
...
+lowerProg (Module _ b:prg) = (++) <$> lowerProg b <*> lowerProg prg
```
Comment (zardyh) [inline src/Core/Lower.hs]: Here, float out the types and make the values into a record.

File: src/Syntax.hs [hunk 5]
```diff
@@ -26,15 +26,26 @@ newtype Typed = Typed Typed deriving Data
 
 data family Var a
 
+class InModule v where
+  inModule :: v -> v -> v
```
Comment (zardyh) [inline src/Syntax.hs]: Can you not use `Monoid` for this - or, well, `Semigroup`?

Comment (SquidDev) [reply src/Syntax.hs]: I'm 90% sure I can replace this with an empty space. I don't think I ended up using it at all.

**Comment by zardyh** [comment]
This looks good, but the fact that modules don't have an explicit computational interpretation makes me wary for the future - I'm thinking of things like first class modules and implicit module parameters, which aiui modules as implemented like this don't let us do.

File: src/Syntax.hs (hunk 5 above)
Comment (zardyh) [reply src/Syntax.hs]: Oh.

File: src/Syntax/Resolve.hs [hunk 6]
```diff
@@ -93,17 +93,7 @@ resolveModule (r:rs) = flip catchError (throwError . flip ArisingFromTop r)
...
-           modP (Name v) = InModule v
-           modP (InModule m v) = InModule m . modP v
-
-           modR _ x@(TgInternal _) = x
-           modR m (TgName v i) = TgName (T.intercalate (T.pack ".") (modParts m [v])) i
-
-           modParts (Name v) xs = v:xs
-           modParts (InModule m v) xs = m:modParts v xs
-
+           modZip name name' v v' = zip (map (name<>) v) (map (name'<>) v')
```
Comment (zardyh) [inline src/Syntax/Resolve.hs]: blessed diff, :+1: squid

**Comment by zardyh** [comment]
Is this done, @SquidDev?

File: src/Parser.y (hunk 1 above)
Comment (zardyh) [reply src/Parser.y]: done

**Comment by zardyh** [comment]
Is this done now?

**Comment by SquidDev** [comment]
Yes?

**Pull request merged**